### PR TITLE
GtkD Event Handling

### PR DESCRIPTION
**Pull request by gnunn1**
This PR addresses #167 by switching the GtkD event handling mechanism to map every delegate to a unique GTK handler on a 1:1 basis. There are a few benefits to this change as follows:

* Delegates can be removed via Signals.handlerDisconnect
* Delegates can be registered against the same signal using different ConnectFlags. Previously, GtkD only registered a signal handler once so you were limited to using a single ConnectFlag per signal

In general this PR is API compatible with the existing GtkD API with the following exceptions:

* The various public onXXXListeners array fields are no longer exposed publically, any code relying on it to remove delegates will need to switch to using Signals.jandlerDisconnect
* The public associatve array connectedSignals has been removed since it is no longer needed.

I have tested these changes with terminix which is a fairly large GTK program. Obviously however there are many areas of GTK it does not use and my testing with terminix is limited. I would recommend care introducing this change with a relatively long gestation period before using it in a public release.

Note I could not test this with building APILookupGLt.txt and APILookupGLd.txt as I don't have those gir files. 

Feedback welcome, if you require any changes prior to accepting this PR just let me know.

### CODE REVIEW

File: wrap/APILookupGObject.txt [hunk 1]
```diff
@@ -322,17 +334,21 @@ code: start
...
+	 * detail strings for the notify signal.
 	 *
 	 * Params:
-	 *     dlg          = The callback.
-	 *     property     = Set this if you only want to receive the signal for a specific property.
-	 *     connectFlags = The behavior of the signal's connection.
+	 *     pspec = the #GParamSpec of the property which changed.
 	 */
-	void addOnNotify(void delegate(ParamSpec, ObjectG) dlg, string property = "", ConnectFlags connectFlags=cast(ConnectFlags)0)
+	gulong addOnNotify(void delegate(ParamSpec, ObjectG) dlg, string property = "", ConnectFlags connectFlags=cast(ConnectFlags)0)
```
Comment (MikeWey) [inline wrap/APILookupGObject.txt]: addOnNotify should keep it's old documentation as it better matches the D version.

Other than that it looks good.

Comment (gnunn1) [inline wrap/APILookupGObject.txt]: Oops, too much copy and paste. I've fixed it, thanks for catching it.

**Comment by gnunn1** [comment]
Looks like I need to tweak the code for 32 bit based on this issue, I'll take care of it later today.

https://github.com/gnunn1/terminix/pull/663#issuecomment-269864178

**Comment by gnunn1** [comment]
I need some advice here, if you compile the code on 32 bit Linux you get the following errors:

```
./GtkD/src/atk/TableT.d(721,33): Error: cannot implicitly convert expression (this.onModelChangedListeners[this.onModelChangedListeners.length - 1u].handlerId) of type ulong to uint
```

In glibtypes.d I can see that gulong is aliased as a uint for anything other then X86_64, so I presume that Windows would have the same issue.  

Should I change this back to ulong instead of gulong?

File: wrap/utils/GtkFunction.d [hunk 2]
```diff
@@ -967,6 +967,35 @@ final class GtkFunction
...
+	
+	string[] getDelegateWrapperDeclaration()
+	{
+		assert(type == GtkFunctionType.Signal);
+
+		string[] buff;
+		buff ~= "protected class " ~ getDelegateWrapperName();
+		buff ~= "{";
+		buff ~= getDelegateDecleration() ~ " dlg;";
+		buff ~= "ulong handlerId;";
```
Comment (MikeWey) [inline wrap/utils/GtkFunction.d]: gulong

File: wrap/utils/GtkFunction.d [hunk 3]
```diff
@@ -967,6 +967,35 @@ final class GtkFunction
...
+		buff ~= "ConnectFlags flags;";
+		buff ~= "this(" ~ getDelegateDecleration() ~ " dlg, ulong handlerId, ConnectFlags flags)";
```
Comment (MikeWey) [inline wrap/utils/GtkFunction.d]: Ditto.

**Comment by gnunn1** [comment]
Well I'm feeling a bit dumb, not sure why I didn't notice that. Thanks for the pointer, I'll clean that up.

**Comment by gnunn1** [comment]
OK I've made those changes so hopefully everything is good to go now. Thanks for all the help Mike.

**Comment by MikeWey** [comment]
We now actually hit the maximum number of symbols in an OMF when compiling everything at one on 32bit windows with dmd. 😄

**Comment by MikeWey** [comment]
I've recently been working on a D implementation of gobject.Closure + a marshaler, which means that now it's possible to pass a D delegate/function directly to Signals.connect.

Since that is now possible i've switched over the addOn* functions over to using this new connect function instead of the wrapper classes. While testing these changes with tilix i've fixed a few bugs that popped up and things seem to be working properly now. Tough a change like this might warrant some more rigorous testing.

**Comment by gnunn1** [comment]
Wow that's a great change, I had a quick look at the new code in the generated folder and it's much cleaner now then what was there previously. I'll update my local instance to use this version and see how it goes.

**Pull request merged**